From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Thu, 15 Oct 2026 07:00:15 -0700
Subject: [PATCH] mscoree: metadata dispenser and GetAssemblyMDImport use the
 rometadata engine (3315)

DECISIONS-2 NET-META. mscoree's own IMetaDataDispenserEx was a table of E_NOTIMPL
stubs. Metadata objects now come from rometadata.dll (the ECMA-335 reader of 3063):
MetaDataDispenser_CreateInstance loads it and returns its dispenser, so OpenScope,
OpenScopeOnMemory, options and the CLR directory work through mscoree too.
GetAssemblyMDImport opens the assembly file for reading and QIs the requested
interface; CoCreateInstance(CLSID_CorMetaDataDispenser[Runtime]) is served directly
by mscoree's class factory instead of needing a Mono runtime registration.

diff --git a/dlls/mscoree/metadata.c b/dlls/mscoree/metadata.c
index e41d610..6890ddc 100644
--- a/dlls/mscoree/metadata.c
+++ b/dlls/mscoree/metadata.c
@@ -1,5 +1,5 @@
 /*
- * IMetaDataDispenserEx - dynamic creation/editing of assemblies
+ * IMetaDataDispenserEx - forwarded to the metadata engine in rometadata.dll
  *
  * Copyright 2010 Vincent Povirk for CodeWeavers
  *
@@ -40,165 +40,51 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL( mscoree );
 
-typedef struct MetaDataDispenser
-{
-    IMetaDataDispenserEx IMetaDataDispenserEx_iface;
-    LONG ref;
-} MetaDataDispenser;
-
-static inline MetaDataDispenser *impl_from_IMetaDataDispenserEx(IMetaDataDispenserEx *iface)
-{
-    return CONTAINING_RECORD(iface, MetaDataDispenser, IMetaDataDispenserEx_iface);
-}
+/* the metadata engine (reader, writer, dispenser) lives in rometadata.dll; mscoree hands out its objects */
+static HRESULT (WINAPI *pMetaDataGetDispenser)(REFCLSID, REFIID, void **);
 
-static HRESULT WINAPI MetaDataDispenser_QueryInterface(IMetaDataDispenserEx* iface,
-    REFIID riid, void **ppvObject)
+static HRESULT load_engine(void)
 {
-    TRACE("%p %s %p\n", iface, debugstr_guid(riid), ppvObject);
-
-    if (IsEqualGUID(riid, &IID_IMetaDataDispenserEx) ||
-        IsEqualGUID(riid, &IID_IMetaDataDispenser) ||
-        IsEqualGUID(riid, &IID_IUnknown))
-    {
-        *ppvObject = iface;
-    }
-    else
-    {
-        FIXME("Unsupported interface %s\n", debugstr_guid(riid));
-        return E_NOINTERFACE;
-    }
-
-    IMetaDataDispenserEx_AddRef( iface );
+    HMODULE module;
 
+    if (pMetaDataGetDispenser) return S_OK;
+    if (!(module = LoadLibraryW(L"rometadata.dll"))) return HRESULT_FROM_WIN32(GetLastError());
+    if (!(pMetaDataGetDispenser = (void *)GetProcAddress(module, "MetaDataGetDispenser")))
+        return HRESULT_FROM_WIN32(ERROR_PROC_NOT_FOUND);
     return S_OK;
 }
 
-static ULONG WINAPI MetaDataDispenser_AddRef(IMetaDataDispenserEx* iface)
-{
-    MetaDataDispenser *This = impl_from_IMetaDataDispenserEx(iface);
-    ULONG ref = InterlockedIncrement(&This->ref);
-
-    TRACE("%p ref=%lu\n", This, ref);
-
-    return ref;
-}
-
-static ULONG WINAPI MetaDataDispenser_Release(IMetaDataDispenserEx* iface)
-{
-    MetaDataDispenser *This = impl_from_IMetaDataDispenserEx(iface);
-    ULONG ref = InterlockedDecrement(&This->ref);
-
-    TRACE("%p ref=%lu\n", This, ref);
-
-    if (ref == 0)
-    {
-        free(This);
-    }
-
-    return ref;
-}
-
-static HRESULT WINAPI MetaDataDispenser_DefineScope(IMetaDataDispenserEx* iface,
-    REFCLSID rclsid, DWORD dwCreateFlags, REFIID riid, IUnknown **ppIUnk)
-{
-    FIXME("%p %s %lx %s %p\n", iface, debugstr_guid(rclsid), dwCreateFlags,
-        debugstr_guid(riid), ppIUnk);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_OpenScope(IMetaDataDispenserEx* iface,
-    LPCWSTR szScope, DWORD dwOpenFlags, REFIID riid, IUnknown **ppIUnk)
-{
-    FIXME("%p %s %lx %s %p\n", iface, debugstr_w(szScope), dwOpenFlags,
-        debugstr_guid(riid), ppIUnk);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_OpenScopeOnMemory(IMetaDataDispenserEx* iface,
-    const void *pData, ULONG cbData, DWORD dwOpenFlags, REFIID riid, IUnknown **ppIUnk)
-{
-    FIXME("%p %p %lu %lx %s %p\n", iface, pData, cbData, dwOpenFlags,
-        debugstr_guid(riid), ppIUnk);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_SetOption(IMetaDataDispenserEx* iface,
-    REFGUID optionid, const VARIANT *value)
-{
-    FIXME("%p %s\n", iface, debugstr_guid(optionid));
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_GetOption(IMetaDataDispenserEx* iface,
-    REFGUID optionid, VARIANT *pvalue)
-{
-    FIXME("%p %s\n", iface, debugstr_guid(optionid));
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_OpenScopeOnITypeInfo(IMetaDataDispenserEx* iface,
-    ITypeInfo *pITI, DWORD dwOpenFlags, REFIID riid, IUnknown **ppIUnk)
-{
-    FIXME("%p %p %lu %s %p\n", iface, pITI, dwOpenFlags, debugstr_guid(riid), ppIUnk);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_GetCORSystemDirectory(IMetaDataDispenserEx* iface,
-    LPWSTR szBuffer, DWORD cchBuffer, DWORD *pchBuffer)
-{
-    FIXME("%p %p %lu %p\n", iface, szBuffer, cchBuffer, pchBuffer);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI MetaDataDispenser_FindAssembly(IMetaDataDispenserEx* iface,
-    LPCWSTR szAppBase, LPCWSTR szPrivateBin, LPCWSTR szGlobalBin, LPCWSTR szAssemblyName,
-    LPWSTR szName, ULONG cchName, ULONG *pcName)
+HRESULT MetaDataDispenser_CreateInstance(IUnknown **ppUnk)
 {
-    FIXME("%p %s %s %s %s %p %lu %p\n", iface, debugstr_w(szAppBase),
-        debugstr_w(szPrivateBin), debugstr_w(szGlobalBin),
-        debugstr_w(szAssemblyName), szName, cchName, pcName);
-    return E_NOTIMPL;
-}
+    HRESULT hr;
 
-static HRESULT WINAPI MetaDataDispenser_FindAssemblyModule(IMetaDataDispenserEx* iface,
-    LPCWSTR szAppBase, LPCWSTR szPrivateBin, LPCWSTR szGlobalBin, LPCWSTR szAssemblyName,
-    LPCWSTR szModuleName, LPWSTR szName, ULONG cchName, ULONG *pcName)
-{
-    FIXME("%p %s %s %s %s %s %p %lu %p\n", iface, debugstr_w(szAppBase),
-        debugstr_w(szPrivateBin), debugstr_w(szGlobalBin), debugstr_w(szAssemblyName),
-        debugstr_w(szModuleName), szName, cchName, pcName);
-    return E_NOTIMPL;
+    if (FAILED(hr = load_engine())) return hr;
+    return pMetaDataGetDispenser(&CLSID_CorMetaDataDispenser, &IID_IUnknown, (void **)ppUnk);
 }
 
-static const struct IMetaDataDispenserExVtbl MetaDataDispenserVtbl =
-{
-    MetaDataDispenser_QueryInterface,
-    MetaDataDispenser_AddRef,
-    MetaDataDispenser_Release,
-    MetaDataDispenser_DefineScope,
-    MetaDataDispenser_OpenScope,
-    MetaDataDispenser_OpenScopeOnMemory,
-    MetaDataDispenser_SetOption,
-    MetaDataDispenser_GetOption,
-    MetaDataDispenser_OpenScopeOnITypeInfo,
-    MetaDataDispenser_GetCORSystemDirectory,
-    MetaDataDispenser_FindAssembly,
-    MetaDataDispenser_FindAssemblyModule
-};
+#ifdef SG_MUTANT_NETMETA_OPEN
+#define SG_OPEN_FLAGS 1
+#else
+#define SG_OPEN_FLAGS 0
+#endif
 
-HRESULT MetaDataDispenser_CreateInstance(IUnknown **ppUnk)
+HRESULT WINAPI GetAssemblyMDImport(LPCWSTR szFileName, REFIID riid, IUnknown **ppIUnk)
 {
-    MetaDataDispenser *This;
+    IMetaDataDispenserEx *dispenser;
+    IUnknown *unk;
+    HRESULT hr;
 
-    This = malloc(sizeof(MetaDataDispenser));
+    TRACE("(%s, %s, %p)\n", debugstr_w(szFileName), debugstr_guid(riid), ppIUnk);
 
-    if (!This)
-        return E_OUTOFMEMORY;
+    if (!szFileName || !riid || !ppIUnk) return E_POINTER;
+    *ppIUnk = NULL;
 
-    This->IMetaDataDispenserEx_iface.lpVtbl = &MetaDataDispenserVtbl;
-    This->ref = 1;
+    if (FAILED(hr = MetaDataDispenser_CreateInstance(&unk))) return hr;
+    hr = IUnknown_QueryInterface(unk, &IID_IMetaDataDispenserEx, (void **)&dispenser);
+    IUnknown_Release(unk);
+    if (FAILED(hr)) return hr;
 
-    *ppUnk = (IUnknown*)&This->IMetaDataDispenserEx_iface;
-
-    return S_OK;
+    hr = IMetaDataDispenserEx_OpenScope(dispenser, szFileName, SG_OPEN_FLAGS /* ofRead */, riid, ppIUnk);
+    IMetaDataDispenserEx_Release(dispenser);
+    return hr;
 }
diff --git a/dlls/mscoree/mscoree_main.c b/dlls/mscoree/mscoree_main.c
index 2b9f78a..94a1262 100644
--- a/dlls/mscoree/mscoree_main.c
+++ b/dlls/mscoree/mscoree_main.c
@@ -513,12 +513,6 @@ HRESULT WINAPI CoInitializeCor(DWORD fFlags)
     return S_OK;
 }
 
-HRESULT WINAPI GetAssemblyMDImport(LPCWSTR szFileName, REFIID riid, IUnknown **ppIUnk)
-{
-    FIXME("(%p %s, %s, %p): stub\n", szFileName, debugstr_w(szFileName), debugstr_guid(riid), *ppIUnk);
-    return ERROR_CALL_NOT_IMPLEMENTED;
-}
-
 HRESULT WINAPI GetVersionFromProcess(HANDLE hProcess, LPWSTR pVersion, DWORD cchBuffer, DWORD *dwLength)
 {
     FIXME("(%p, %p, %ld, %p): stub\n", hProcess, pVersion, cchBuffer, dwLength);
@@ -705,6 +699,11 @@ HRESULT WINAPI CreateInterface(REFCLSID clsid, REFIID riid, LPVOID *ppInterface)
     return CLRCreateInstance(clsid, riid, ppInterface);
 }
 
+static HRESULT create_dispenser(REFCLSID clsid, LPVOID *ppObj)
+{
+    return MetaDataDispenser_CreateInstance((IUnknown **)ppObj);
+}
+
 HRESULT WINAPI DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID* ppv)
 {
     mscorecf *This;
@@ -718,7 +717,12 @@ HRESULT WINAPI DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID* ppv)
     This = malloc(sizeof(mscorecf));
 
     This->IClassFactory_iface.lpVtbl = &mscorecf_vtbl;
-    This->pfnCreateInstance = create_monodata;
+#ifndef SG_MUTANT_NETMETA_CLASS
+    if (IsEqualGUID(rclsid, &CLSID_CorMetaDataDispenser) || IsEqualGUID(rclsid, &CLSID_CorMetaDataDispenserRuntime))
+        This->pfnCreateInstance = create_dispenser;
+    else
+#endif
+        This->pfnCreateInstance = create_monodata;
     This->ref = 1;
     This->clsid = *rclsid;
 
